ext/opentelemetry-ext-celery/tests/test_utils.py

```diff
@@ -0,0 +1,210 @@
+            span.attributes["celery.delivery_info"], str({"eager": True})
+        )
+        self.assertEqual(span.attributes.get("celery.eta"), "soon")
+        self.assertEqual(span.attributes.get("celery.expires"), "later")
+        self.assertEqual(span.attributes.get("celery.hostname"), "localhost")
+
+        self.assertEqual(span.attributes.get("celery.reply_to"), "44b7f305")
+        self.assertEqual(span.attributes.get("celery.retries"), 4)
+        self.assertEqual(
+            span.attributes.get("celery.timelimit"), ("now", "later")
+        )
+        self.assertNotIn("custom_meta", span.attributes)
+
+    def test_set_attributes_from_context_empty_keys(self):
+        # it should not extract empty keys
+        context = {
+            "correlation_id": None,
+            "exchange": "",
+            "timelimit": (None, None),
+            "retries": 0,
+        }
+
+        span = trace.Span("name", mock.Mock(spec=trace_api.SpanContext))
+        utils.set_attributes_from_context(span, context)
+
+        self.assertEqual(len(span.attributes), 0)
+        # edge case: `timelimit` can also be a list of None values
+        context = {
+            "timelimit": [None, None],
+        }
+
+        utils.set_attributes_from_context(span, context)
+
+        self.assertEqual(len(span.attributes), 0)
+
+    def test_span_propagation(self):
+        # ensure spans getter and setter works properly
+        @self.app.task
+        def fn_task():
+            return 42
+
+        # propagate and retrieve a Span
+        task_id = "7c6731af-9533-40c3-83a9-25b58f0d837f"
+        span = trace.Span("name", mock.Mock(spec=trace_api.SpanContext))
+        utils.attach_span(fn_task, task_id, span)
+        span_after = utils.retrieve_span(fn_task, task_id)
+        self.assertIs(span, span_after)
+
+    def test_span_delete(self):
+        # ensure the helper removes properly a propagated Span
+        @self.app.task
+        def fn_task():
+            return 42
+
+        # propagate a Span
+        task_id = "7c6731af-9533-40c3-83a9-25b58f0d837f"
+        span = trace.Span("name", mock.Mock(spec=trace_api.SpanContext))
+        utils.attach_span(fn_task, task_id, span)
+        # delete the Span
+        utils.detach_span(fn_task, task_id)
+        self.assertEqual(utils.retrieve_span(fn_task, task_id), (None, None))
+
+    def test_span_delete_empty(self):
+        # ensure detach_span doesn't raise an exception if span is not present
+        @self.app.task
+        def fn_task():
+            return 42
+
+        # delete the Span
+        exception = None
+        task_id = "7c6731af-9533-40c3-83a9-25b58f0d837f"
+        try:
+            utils.detach_span(fn_task, task_id)
+            self.assertEqual(
+                utils.retrieve_span(fn_task, task_id), (None, None)
+            )
+        except Exception as ex:  # pylint: disable=broad-except
+            exception = ex
+        self.assertIsNone(exception)
```

Comment:
`self.fail` should be used instead if the purpose of the test is to assert no exception was raised.

Reply:
Thanks, updated.